apollo-core/src/main/java/com/ctrip/framework/apollo/core/utils/DeferredLogCache.java

```diff
@@ -0,0 +1,168 @@
+ * @since 2021/5/11
+ */
+final class DeferredLogCache {
+
+  public static final int MAX_LOG_SIZE = 1000;
+  private static final AtomicInteger LOG_INDEX = new AtomicInteger(0);
+  private static final Cache<Integer, Line> LOG_CACHE = CacheBuilder.newBuilder()
+      .maximumSize(MAX_LOG_SIZE)
+      .expireAfterWrite(1, TimeUnit.MINUTES)
+      .build();
+
+  private DeferredLogCache() {
+  }
+
+
+  public static void debug(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.DEBUG, message, objects, throwable);
+  }
+
+  public static void info(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.INFO, message, objects, throwable);
+  }
+
+  public static void warn(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.WARN, message, objects, throwable);
+  }
+
+  public static void error(Logger logger, String message, Throwable throwable, Object... objects) {
+    add(logger, Level.ERROR, message, objects, throwable);
+  }
+
+  private static void add(Logger logger, Level level, String message, Object[] objects,
+      Throwable throwable) {
+    Line logLine = new Line(level, message, objects, throwable, logger);
+    LOG_CACHE.put(LOG_INDEX.incrementAndGet(), logLine);
+  }
+
+  public static void replayTo() {
+    for (int i = 1; i <= LOG_INDEX.get(); i++) {
+      Line logLine = LOG_CACHE.getIfPresent(i);
+      assert logLine != null;
+      Logger logger = logLine.getLogger();
+      Level level = logLine.getLevel();
+      String message = logLine.getMessage();
+      Object[] objects = logLine.getObjects();
+      Throwable throwable = logLine.getThrowable();
+      logTo(logger, level, message, objects, throwable);
+    }
+    clear();
+  }
+
+  public static void clear() {
+    LOG_CACHE.invalidateAll();
+    LOG_INDEX.set(0);
+  }
+
+  public static long logSize() {
+    return LOG_CACHE.size();
+  }
+
+  static void logTo(Logger logger, Level level, String message, Object[] objects,
+      Throwable throwable) {
+    switch (level) {
+      case DEBUG:
+        if (throwable != null) {
+          logger.debug(message, throwable);
+        } else {
+          logger.debug(message, objects);
+        }
+        return;
+      case INFO:
+        if (throwable != null) {
+          logger.info(message, throwable);
+        } else {
+          logger.info(message, objects);
+        }
+        return;
+      case WARN:
+        if (throwable != null) {
+          logger.warn(message, throwable);
+        } else {
+          logger.warn(message, objects);
+        }
+        return;
+      case ERROR:
+        if (throwable != null) {
+          logger.error(message, throwable);
+        } else {
+          logger.error(message, objects);
+        }
+        break;
+      default:
+        throw new IllegalStateException("Unexpected value: " + level);
+    }
+  }
+
+  static class Line {
```

Comment:
this class could be private
```suggestion
  privarte static class Line {
```